### DISCUSSION TITLE
how to make the corners of a the main window rounded when it is overrideredirected

### DISCUSSION THREAD

**Discussion by Zingzy** [TomSchimansky/CustomTkinter#1302] [Q&A]
How to make the corners of the main windows when the overrideredirect is set to true?

**Comment by Akascape** [comment]
@Zingzy here is an example of round edges on main window with overrideredirect:
```python
import customtkinter, sys

root = customtkinter.CTk()
root.grid_columnconfigure(0, weight=1)
root.grid_rowconfigure(0, weight=1)

root.overrideredirect(True) # Removes the header
corner = 30 # corner radius

# have to add some transparency for bg_color
if sys.platform.startswith("win"):
    # for windows
    transparent_color = '#000001'
    root.attributes("-transparentcolor", transparent_color)
elif sys.platform.startswith("darwin"):
    # for mac
    transparent_color = 'systemTransparent'
    root.attributes("-transparent", True)
else:
    # transparency not available for linux based os, so no round corners for them
    corner = 0
    
main_frame = customtkinter.CTkFrame(root, corner_radius=corner, bg_color=transparent_color)
main_frame.grid(sticky="nswe")
main_frame.grid_columnconfigure(0, weight=1)
main_frame.grid_rowconfigure(0, weight=1)

root.mainloop()
```

If you are on windows, then there is ready-made **customizable** round main window I made: https://github.com/TomSchimansky/CustomTkinter/discussions/736

**Reply by Zingzy** [reply]
I tried this trick, but is there anything else we could do?

**Reply by Akascape** [reply]
@Zingzy What else? Please let me what exactly you are making?

Don't forget that customtkinter is just build in top of tkinter. So the core functions are the same as tkinter.
There is no difference in the CTk() and Tk() window, that is why there is no corner_radius option for that. It will always appear as square if you use overrideredirect.

Moreover, the corner roundess of the main window
 (CTk or Tk) is dependent on the **operating system** only. Windows 11, ubuntu, mac have round corners by default. Windows 10 lacks this feature so it is impossible to do that in the core. **We can't do anything with that**

But when it comes to customtkinter, CTkFrame has 2 portions, **bg_color and fg_color**, separated by the **round** edges. We made the bg_color transparent by using tkinter window attributes which gives us the round corner transparency effect. Other than that, there is no different trick.

I have made a full customizable window for those who are using windows 10 which includes all the basic things of the main window.

**Reply by Zingzy** [reply]
Ok so I am using windows 11 and I just found another trick, by using `self.wm_attributes("-toolwindow", True)` I could hide the other buttons but the only problem is that it is creating another ugly close button... Is there any way to modify that button? Also when I use this command the window remains rounded

**Reply by Akascape** [reply]
Yes, you can do it in windows only.
No other tweak can be used for the header, it is controlled by os.

**Reply by Akascape** [reply]
@Zingzy 
> I am using windows 11 and I just found another trick, by using self.wm_attributes("-toolwindow", True) I could hide the other buttons but the only problem is that it is creating another ugly close button...

I just found the proper method to make the toolwindow without the ugly looking close button.

```python
self.resizable(width=False, height=False)
self.transient(self.master)
```

**Reply by Zingzy** [reply]
> @Zingzy
> 
> > I am using windows 11 and I just found another trick, by using self.wm_attributes("-toolwindow", True) I could hide the other buttons but the only problem is that it is creating another ugly close button...
> 
> I just found the proper method to make the toolwindow without the ugly looking close button.
> 
> ```python
> self.resizable(width=False, height=False)
> self.transient(self.master)
> ```

hey thanks for the tip but I dont get it how to use this code, can you please show a full short tkitner code with this for the context?

**Reply by Akascape** [reply]
@Zingzy 
Here is a small example:
```python
import customtkinter

root = customtkinter.CTkToplevel() # use only with toplevel windows
root.resizable(width=False, height=False)
root.transient(root.master)
root.master.attributes("-alpha", 0)
root.mainloop()
```
![](https://user-images.githubusercontent.com/89206401/229334130-86f9dfe4-5b7c-487a-b768-11ae0e888f84.jpg)

**Reply by Zingzy** [reply]
> @Zingzy Here is a small example:
> 
> ```python
> import customtkinter
> 
> root = customtkinter.CTkToplevel() # use only with toplevel windows
> root.resizable(width=False, height=False)
> root.transient(root.master)
> root.mainloop()
> ```
> 
> ![](https://user-images.githubusercontent.com/89206401/229334130-86f9dfe4-5b7c-487a-b768-11ae0e888f84.jpg)

Ok great but there is only one problems, this creates an additional window of CTk 
If there is a way to remove the main window?

**Reply by Akascape** [reply]
@Zingzy Ok, I got the fix.

> Ok great but there is only one problems, this creates an additional window of CTk
If there is a way to remove the main window?

You can add `root.master.attributes("-alpha", 0)` in order to hide the additional window. I have modified the example.
